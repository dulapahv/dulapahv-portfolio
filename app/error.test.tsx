import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import Error from './error';

// Mock next/link
vi.mock('next/link', () => ({
  default: ({
    children,
    href,
    className
  }: {
    children: React.ReactNode;
    href: string;
    className?: string;
  }) => (
    <a href={href} className={className}>
      {children}
    </a>
  )
}));

// Mock Footer component
vi.mock('@/components/footer', () => ({
  default: () => <footer>Footer</footer>
}));

describe('Error Page', () => {
  const mockError = {
    name: 'TestError',
    message: 'This is a test error',
    digest: 'abc123',
    cause: 'Test cause'
  } as Error & { digest?: string };

  const mockReset = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
    // Mock console.error to avoid cluttering test output
    vi.spyOn(console, 'error').mockImplementation(() => {});
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('should render error heading', () => {
    render(<Error error={mockError} reset={mockReset} />);
    expect(screen.getByRole('heading', { name: 'Oops! Something Went Wrong' })).toBeInTheDocument();
  });

  it('should render error message with contact link', () => {
    render(<Error error={mockError} reset={mockReset} />);
    expect(screen.getByText(/An error occurred while rendering this page/i)).toBeInTheDocument();

    const contactLink = screen.getByRole('link', { name: /contact me/i });
    expect(contactLink).toBeInTheDocument();
    expect(contactLink).toHaveAttribute('href', expect.stringContaining('/contact'));
  });

  it('should render try again button', () => {
    render(<Error error={mockError} reset={mockReset} />);
    const tryAgainButton = screen.getByRole('button', { name: /try again/i });
    expect(tryAgainButton).toBeInTheDocument();
  });

  it('should call reset function when try again button is clicked', () => {
    render(<Error error={mockError} reset={mockReset} />);

    const tryAgainButton = screen.getByRole('button', { name: /try again/i });
    fireEvent.click(tryAgainButton);

    expect(mockReset).toHaveBeenCalledTimes(1);
  });

  it('should display error details', () => {
    render(<Error error={mockError} reset={mockReset} />);

    expect(screen.getByText(/Details:/i)).toBeInTheDocument();
    expect(screen.getByText(/Name: TestError/i)).toBeInTheDocument();
    expect(screen.getByText(/Message: This is a test error/i)).toBeInTheDocument();
    expect(screen.getByText(/Cause: Test cause/i)).toBeInTheDocument();
    expect(screen.getByText(/Digest: abc123/i)).toBeInTheDocument();
  });

  it('should log error to console on mount', () => {
    const consoleErrorSpy = vi.spyOn(console, 'error');
    render(<Error error={mockError} reset={mockReset} />);

    expect(consoleErrorSpy).toHaveBeenCalledWith(mockError);
  });

  it('should handle error without optional fields', () => {
    const minimalError = {
      name: '',
      message: ''
    } as Error & { digest?: string };

    render(<Error error={minimalError} reset={mockReset} />);

    expect(screen.getByText(/Name: N\/A/i)).toBeInTheDocument();
    expect(screen.getByText(/Message: N\/A/i)).toBeInTheDocument();
    expect(screen.getByText(/Cause: N\/A/i)).toBeInTheDocument();
    expect(screen.getByText(/Digest: N\/A/i)).toBeInTheDocument();
  });

  it('should render Footer component', () => {
    render(<Error error={mockError} reset={mockReset} />);
    expect(screen.getByText('Footer')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { container } = render(<Error error={mockError} reset={mockReset} />);
    expect(container).toMatchSnapshot();
  });
});
