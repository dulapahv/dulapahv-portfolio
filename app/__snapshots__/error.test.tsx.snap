// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Error Page > should match snapshot 1`] = `
<div>
  <main
    class="space-y-4"
  >
    <header>
      <h1
        class="text-3xl font-semibold"
      >
        Oops! Something Went Wrong
      </h1>
    </header>
    <p
      class="text-foreground-muted"
    >
      An error occurred while rendering this page. Please try again later or
       
      <a
        class="text-mirai-red underline underline-offset-2 hover:decoration-2"
        href="/contact?message=Please%20describe%20what%20you%20were%20doing%20when%20this%20error%20occurred%3A%0A%0A%0A%0APlease%20do%20not%20remove%20the%20details%20below%2C%20they%20help%20me%20identify%20and%20fix%20the%20issue%20faster.%0A%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%0AError%20Details%3A%0ATimestamp%3A%2009%2F01%2F2007%2C%2009%3A41%3A00%20(2007-01-09T09%3A41%3A00.000Z)%0APath%3A%20%2F%0AName%3A%20TestError%0AMessage%3A%20This%20is%20a%20test%20error%0ACause%3A%20Test%20cause%0ADigest%3A%20abc123%0A%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D"
      >
        contact me
      </a>
       
      if the issue persists.
    </p>
    <button
      class="bg-mirai-red flex w-fit cursor-pointer items-center justify-center gap-2 rounded-md px-3 py-2 text-sm font-medium text-white select-none hover:bg-mirai-red/90 transition-colors hover:shadow-md active:scale-[0.98] active:transition-transform! disabled:cursor-not-allowed disabled:opacity-50 disabled:hover:shadow-none"
    >
      <svg
        aria-hidden="true"
        class="size-4.5 shrink-0"
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M224,48V96a8,8,0,0,1-8,8H168a8,8,0,0,1,0-16h28.69L182.06,73.37a79.56,79.56,0,0,0-56.13-23.43h-.45A79.52,79.52,0,0,0,69.59,72.71,8,8,0,0,1,58.41,61.27a96,96,0,0,1,135,.79L208,76.69V48a8,8,0,0,1,16,0ZM186.41,183.29a80,80,0,0,1-112.47-.66L59.31,168H88a8,8,0,0,0,0-16H40a8,8,0,0,0-8,8v48a8,8,0,0,0,16,0V179.31l14.63,14.63A95.43,95.43,0,0,0,130,222.06h.53a95.36,95.36,0,0,0,67.07-27.33,8,8,0,0,0-11.18-11.44Z"
        />
      </svg>
      <span>
        Try Again
      </span>
    </button>
  </main>
  <div
    class="text-foreground-muted mt-8 border-t border-gray-200 pt-6 dark:border-gray-800"
  >
    <p
      class="mb-2 font-semibold"
    >
      Details:
    </p>
    <code
      class="block text-sm break-all whitespace-pre-wrap sm:text-base"
    >
      Timestamp: 09/01/2007, 09:41:00 (2007-01-09T09:41:00.000Z)
      <br />
      Path: /
      <br />
      Name: TestError
      <br />
      Message: This is a test error
      <br />
      Cause: Test cause
      <br />
      Digest: abc123
    </code>
  </div>
  <footer>
    Footer
  </footer>
</div>
`;
